pkg/innerring/processors/settlement/calls.go

```diff
@@ -38,23 +46,87 @@ func (p *Processor) HandleBasicIncomeEvent(e event.Event) {
 
 	l.Info("start basic income distribution...", zap.Int("numberOfContainers", len(cnrs)))
 
-	p.sendPaymentTXs(l, epoch, cnrs)
+	p.sendPaymentTXs(ctx, l, epoch, cnrs)
 
 	l.Debug("finished basic income distribution")
 }
 
-func (p *Processor) sendPaymentTXs(l *zap.Logger, epoch uint64, cnrs []cid.ID) {
-	var wg errgroup.Group
+type nrInfo struct {
+	cID   cid.ID
+	mainH util.Uint256
+	vub   uint32
+}
+
+func (p *Processor) sendPaymentTXs(ctx context.Context, l *zap.Logger, epoch uint64, cnrs []cid.ID) {
+	var (
+		wg errgroup.Group
+
+		m   sync.Mutex
+		res []nrInfo
+	)
 	wg.SetLimit(parallelFactor)
 	for _, cID := range cnrs {
+		select {
+		case <-ctx.Done():
+			_ = wg.Wait()
+			l.Info("interupted container payments settlement", zap.Error(context.Cause(ctx)))
+		default:
+		}
+
 		wg.Go(func() error {
-			err := p.balanceClient.SettleContainerPayment(epoch, cID)
+			var (
+				mH         util.Uint256
+				vub        uint32
+				expBackoff = backoff.NewExponentialBackOff()
+			)
+			err := backoff.RetryNotify(func() error {
+				var err error
+				mH, _, vub, err = p.balanceClient.SettleContainerPayment(ctx, epoch, cID)
+				if err != nil {
+					if errors.Is(err, neorpc.ErrMempoolCapReached) {
+						return err
+					}
+					return backoff.Permanent(err)
+				}
+				return nil
+			},
+				expBackoff,
+				func(err error, d time.Duration) {
+					l.Info("retrying sending payment notary request due to error", zap.Stringer("cID", cID),
+						zap.Duration("retry-after", d), zap.Error(err))
+				})
 			if err != nil {
 				l.Error("could not send payment transaction", zap.Stringer("cID", cID), zap.Error(err))
 				return nil
 			}
 
-			l.Debug("successfully sent transfer", zap.Stringer("cID", cID))
+			l.Debug("successfully sent payment notary request", zap.Stringer("cID", cID))
+
+			m.Lock()
+			res = append(res, nrInfo{cID: cID, mainH: mH, vub: vub})
+			m.Unlock()
+
+			return nil
+		})
+	}
+	_ = wg.Wait()
```

Comment:
Then this doesn't make much sense to me now. It's needlessly complex, the idea was to get rid of worker groups completely, they're just not needed (sending transactions is fast, can be done in a single routine). But whatever, we can have push/await in 20 threads as well.